File: examples/tlv++/src/quick_start.cpp
#include <array>
#include <cstring>

#include "tlv++/tlv.hpp"
#include "tlv++/builtins/fixed/fixed_format.hpp"

int main() {
    using format = tlv::fixed_format<1, 1, TLV_BYTE_ORDER_BIG_ENDIAN>;

    const tlv::tag_t               tag = TLV_TAG(0x01);
    const std::array<tlv::byte, 3> value = {
        static_cast<tlv::byte>(0xAA), static_cast<tlv::byte>(0xBB), static_cast<tlv::byte>(0xCC)};
    std::array<tlv::byte, 5> buffer{};
    tlv::writer              writer(buffer.data(), buffer.size(), format::writer());

    if (!writer.write(tag, tlv::bytes(value.data(), value.size()))) return 1;

    // entry.value borrows buffer; keep it alive while using the entry.
    tlv::reader reader(tlv::bytes(buffer.data(), writer.size()), format::reader());
    auto        entry = reader.next();
    if (!entry || !reader.at_end()) return 1;

    if (entry->tag.size != 1 || entry->tag.data[0] != 0x01) return 1;
    if (entry->value.size() != value.size() ||
        std::memcmp(entry->value.data(), value.data(), value.size()) != 0)
        return 1;
    return 0;
}

File: bindings/lua/examples/quick_start.lua
--[[
The simplest possible read: one element encoded with the configurable
fixed-width format (one tag byte, one length byte), decoded with a Reader.
This binding covers Reader, Entry and Tag only (see issue #298); there is no
Lua Writer yet, so unlike the C, C++, Rust and Python quick_start examples,
the encoded bytes below are written out by hand instead of with a Writer.
See parse.lua for a nested BER document and error handling.

Run with `lua examples/quick_start.lua` from `bindings/lua`, after building
the native module (see ../README.md or docs/development/lua.md).
]]

local opentlv = require("opentlv")

local function hex(bytes)
    local parts = {}
    for i = 1, #bytes do
        parts[i] = string.format("%02X", string.byte(bytes, i))
    end
    return table.concat(parts, " ")
end

-- tag 0x01, length 0x03, value 0xAA 0xBB 0xCC.
local encoded = string.char(0x01, 0x03, 0xAA, 0xBB, 0xCC)
print(string.format("reading %d bytes: %s", #encoded, hex(encoded)))

local format = opentlv.formats.fixed(1, 1, "big")
local reader = opentlv.reader(encoded, format)
local entry = reader:next()
assert(entry ~= nil)
assert(reader:next() == nil)

assert(entry.tag == string.char(0x01))
assert(entry.value == string.char(0xAA, 0xBB, 0xCC))
print(string.format("read tag %s value %s", hex(entry.tag), hex(entry.value)))
print("opentlv version: " .. opentlv.version())
